core/src/main/java/org/apache/lucene/search/uhighlight/CustomUnifiedHighlighter.java

```diff
@@ -0,0 +1,160 @@
+        this.returnNonHighlightedSnippets = returnNonHighlightedSnippets;
+        this.requireFieldMatch = requireFieldMatch;
+    }
+
+    /**
+     * Highlights terms extracted from the provided query within the content of the provided field name
+     */
+    public Snippet[] highlightField(String field, Query query, int docId, int maxPassages) throws IOException {
+        Map<String, Object[]> fieldsAsObjects = super.highlightFieldsAsObjects(new String[]{field}, query,
+            new int[]{docId}, new int[]{maxPassages});
+        Object[] snippetObjects = fieldsAsObjects.get(field);
+        if (snippetObjects != null) {
+            //one single document at a time
+            assert snippetObjects.length == 1;
+            Object snippetObject = snippetObjects[0];
+            if (snippetObject != null && snippetObject instanceof Snippet[]) {
+                return (Snippet[]) snippetObject;
+            }
+        }
+        return EMPTY_SNIPPET;
+    }
+
+    @Override
+    protected List<CharSequence[]> loadFieldValues(String[] fields,
+                                                   DocIdSetIterator docIter,
+                                                   int cacheCharsThreshold) throws IOException {
+        //we only highlight one field, one document at a time
+        return Collections.singletonList(new String[]{fieldValue});
+    }
+
+    @Override
+    protected BreakIterator getBreakIterator(String field) {
+        if (breakIterator != null) {
+            return breakIterator;
+        }
+        return super.getBreakIterator(field);
+    }
+
+    @Override
+    protected PassageFormatter getFormatter(String field) {
+        return passageFormatter;
+    }
+
+    @Override
+    protected OffsetSource getOptimizedOffsetSource(String field,
+                                                    BytesRef[] terms,
+                                                    PhraseHelper phraseHelper,
+                                                    CharacterRunAutomaton[] automata) {
+        if (offsetSource != null) {
+            return offsetSource;
+        }
+        return super.getOptimizedOffsetSource(field, terms, phraseHelper, automata);
+    }
+
+    @Override
+    protected int getMaxNoHighlightPassages(String field) {
+        if (returnNonHighlightedSnippets) {
+            return 1;
+        }
+        return 0;
+    }
+
+    @Override
+    protected BytesRef[] filterExtractedTerms(String field, SortedSet<Term> queryTerms) {
+        if (requireFieldMatch == false) {
+            SortedSet<BytesRef> terms = new TreeSet<>();
+            for (Term term : queryTerms) {
+                terms.add(term.bytes());
+            }
+            return terms.toArray(new BytesRef[terms.size()]);
+        }
+        return super.filterExtractedTerms(field, queryTerms);
+    }
+
+    @Override
+    protected PhraseHelper getPhraseHelper(String field, Query query, EnumSet<HighlightFlag> highlightFlags) {
```

Comment:
The PhraseHelper is likely to become package-private in a future LUCENE JIRA issue.  Why do you need access to it?